### PR TITLE
Import registrations from a CSV file

### PR DESCRIPTION
**Pull request by jonatanklosko**
This allows third-party registration platforms to import their data into the WCA database, so that WCIF includes the registration regardless WCA registration being used. There are other nice side effects, e.g. once imported, the registrations are visible on the competitors page, and well as psych sheets.

The importing was quite tricky due to numerous different scenarios. The specs describe all the possible scenarios, and here's a flowchart for the reference:
```
Registrant has WCA ID
  Yes ->
    WCA ID is connected to a user ->
      Yes ->
        Is dummy user ->
          Yes ->
            There is an account with the registrant's email.
              Yes ->
                The account has WCA ID
                  Yes -> Return an error (account has registrant's email, but different WCA ID, not much we can do about it).
                  No -> Merge the dummy user with this account (i.e. set the account's WCA ID and remove the dummy user) and register it.
              No -> Promote it to a locked account and register it.
          No -> Register this user.
      No -> Create a locked account with the WCA ID (confirmed) and register it.
  No ->
    There is a user with registrant's email
      Yes -> Register this user.
      No -> Create a locked account without WCA ID and register it.
```

The above introduces a new term of a **locked account**. That's simply a non-dummy user with empty `encrypted_password` (which effectively means that it's a normal user that cannot sign in, until he resets his password via email).

Long story short: we reuse existing users whenever there is a reasonable connection between the given registration data and that user. Otherwise we create a locked account and send an email to inform the registrant about this and let him know how to unlock the new account.

### CODE REVIEW

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -98,6 +108,101 @@ def export
...
+            if email_user.wca_id.present?
+              raise "There is already a user with email #{registration[:email]}"\
+                    ", but it has WCA ID of #{email_user.wca_id} instead of #{registration[:wca_id]}."
+            else
+              email_user.update!(wca_id: registration[:wca_id]) # User hooks will also remove the dummy user account.
+              email_user
+            end
+          else
+            user.skip_reconfirmation!
+            user.update!(email: registration[:email]) # TODO: add dummy_account: false once #3722 gets merged.
```
Comment (jonatanklosko) [inline WcaOnRails/app/controllers/registrations_controller.rb]: Note: as per the `TODO` comment, related to #3722.

**Comment by jonatanklosko** [comment]
The current UI:
image

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 1 above)
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Updated!

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -98,6 +108,101 @@ def export
...
+            if email_user.wca_id.present?
+              raise "There is already a user with email #{registration[:email]}"\
+                    ", but it has WCA ID of #{email_user.wca_id} instead of #{registration[:wca_id]}."
+            else
+              email_user.update!(wca_id: registration[:wca_id]) # User hooks will also remove the dummy user account.
+              email_user
+            end
+          else
+            user.skip_reconfirmation!
+            user.update!(dummy_account: false, email: registration[:email])
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Yeah, good catch!

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -98,6 +108,101 @@ def export
...
+    registrations = CSV.read(file.path, headers: true, header_converters: :symbol, skip_blanks: true, converters: ->(string) { string&.strip })
+                       .map(&:to_hash)
+                       .reject { |registration| registration.values.all?(&:nil?) }
+                       .select { |registration| registration[:status] == "a" }
+    new_locked_users = []
+    ActiveRecord::Base.transaction do
+      registrations.each do |registration|
+        user = user_for_registration!(registration, new_locked_users)
+        registration_competition_events_attributes = competition.competition_events.map do |competition_event|
+          registration[competition_event.event_id.to_sym] == "1" ? { competition_event_id: competition_event.id } : nil
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Sounds good to me, what about an error message like this:
image

[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: I love it!

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -98,6 +108,101 @@ def export
...
+        end.compact
+        competition.registrations.create!(
+          user_id: user.id,
+          accepted_at: Time.now,
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Hmm, I think there is more to FIFO checking than just these two timestamps. If registrations get approved once a payment is done, then we would need to know the payment time. On the other hand a lot of platforms don't even have that timestamp, because payments are done via a bank transfer instead of an external provider embedded into one's site. So organizers just approve registrations when the the based on bank transfer dates.

Btw. thinking about this, I've realized that we should probably be checking if the number of approved registrations to be imported doesn't exceed registration limit, right?

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 5]
```diff
@@ -98,6 +108,101 @@ def export
...
+    User.new(
+      name: registration[:name],
+      email: registration[:email],
+      wca_id: registration[:wca_id],
+      country_iso2: Country.c_find(registration[:country]).iso2,
+      gender: registration[:gender],
+      dob: registration[:birth_date],
+      encrypted_password: "",
+    ).tap do |user|
+      user.define_singleton_method(:password_required?) { false }
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: [`define_singleton_method`](https://ruby-doc.org/core-2.5.1/Object.html#method-i-define_singleton_method) simply let's you add/override a method for the given object, with no impact on other objects from its class.` I found it [here](https://stackoverflow.com/a/45589123). The comment above that suggests to do it the other way, I just prefer not to add `attr_accessor`s for a one-time usage, if possible. I can comment this line to point to the SO answer, or use the other workaround, what do you prefer?

The `create_locked_account!` method accepts a very specific hash (a CSV row), but I'm ok with moving it to the `User` model. `User.new_locked_account` you suggested implies a new unsaved instance, I'd rather stick with `create_` to keep the saving stuff together (i.e. `define_singleton_method`, `skip_confirmation!`, `save!`).

File: WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<p>
+  Your registration for <%= @competition.name %> has been imported to the WCA website
+  and as a result a locked account has been created for you.
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb]: 1. I don't think so. Normally when sending an email to registrant, we would check `@registration.user.locale`, but for a newly created locked account we don't have that info. We could take a guess based on their country, but I thought it's alright to keep it in English. Any preference here?
2. Agreed, I will update the wording.
    - concerning side note: initially I planned to just include an activation URL, but soon I realized the tokens have expiration time (and we definitely don't want an exception in this case). So I figured reusing the existing flow is just fine, and more secure than alternatives. Honestly I think most users don't keep their emails open, so in most cases we didn't *just* email them ^^
3. Yeah, I thought of an alternative where you get a clear message during account creation instead. It has one caveat though, if someone (let's assume with WCA ID) registered for a competition using one email, and signs up with another, then not only does he unnecessarily create a duplicated account, but also he cannot claim his WCA ID from the new account, as it's already claimed to the locked account. I don't have strong preference on that, email just seems fine. We can implement both if you like, i.e. leave email notifications, but also improve the error message if someone signs up with an email belonging to a locked account. What do you prefer?

File: WcaOnRails/config/locales/en.yml [hunk 7]
```diff
@@ -952,6 +951,12 @@ en:
     preferred_events_prompt_html: "To speed up registration in the future, you can set your preferred events %{link}."
     preferred_events_populated_html: "We autopopulated this with your preferred events, which you can edit %{link}."
     registration_requirements: "Registration requirements for the competition:"
+    import:
+      title: "Import registrations for %{comp}"
+      warning: "This action is not reversible, make sure to do this once the registration is closed."
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/config/locales/en.yml]: I went for `This action is not reversible and should be done only once, so don't do this until after registration is closed.` for now.

File: WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb]: I've updated the wording like so:
image
How does it sound?

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 8]
```diff
@@ -98,6 +108,101 @@ def export
...
+            user.skip_reconfirmation!
+            user.update!(dummy_account: false, email: registration[:email])
+            user
+          end
+        else
+          user # Use this account.
+        end
+      else
+        # Create a locked account with confirmed WCA ID.
+        create_locked_account!(registration).tap do |locked_user|
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Ah, good question, at this point the error would be `Validation failed: WCA ID not found`, but this doesn't help in identifying the problematic WCA ID. I updated it like so:
image

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 9]
```diff
@@ -98,6 +108,101 @@ def export
...
+    new_locked_users.each do |user|
+      RegistrationsMailer.notify_registrant_of_locked_account_creation(user, competition).deliver_later
+    end
+    redirect_to competition_registrations_import_url(competition)
+  rescue StandardError => error
+    flash[:danger] = error.to_s
+    redirect_to competition_registrations_import_url(competition)
+  end
+
+  private def user_for_registration!(registration, new_locked_users)
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Definitely! I don't even like the current patter, just didn't come up with anything better at the time =)

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: That's how the mentioned error looks:
image

File: WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/views/registrations_mailer/notify_registrant_of_locked_account_creation.html.erb]: > but for a newly created locked account we don't have that info

Oh, right, duh! This seems fine to me =)

> registered for a competition using one email, and signs up with another, then not only does he unnecessarily create a duplicated account, but also he cannot claim his WCA ID from the new account, as it's already claimed to the locked account

Ah, good point! I *think* that if they try to claim their WCA ID while creating the account, then they will get a message about that WCA ID already being assigned to a different account, but it would not be easy for them to figure out which account that is.

I don't know that this email will actually prevent that situation. We already get people who don't really understand email and accounts and passwords creating lots of duplicate accounts. =( However, I'm sold on keeping the email! I also like the idea of improving the signup flow anyways, but we can take care of that later.

> I've updated the wording like so:

I like it! I would like to add a sentence or two about the purpose of these accounts, though. Maybe something like this?

> With this account, you can add a picture to your WCA results profile page, and register for other competitions around the world.

Feel free to reword. I just feel pretty bad about emailing someone about something without telling them *why* they should care about it.

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Yeah, you're totally right about FIFO information. Let's not worry about this right now, I just wanted to start the discussion.

Checking the registration limit seems like a good idea. Thanks for doing that!

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Love it, thanks!

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 10]
```diff
@@ -126,12 +126,21 @@ def do_import
                        .map(&:to_hash)
                        .reject { |registration| registration.values.all?(&:nil?) }
                        .select { |registration| registration[:status] == "a" }
+    if competition.competitor_limit_enabled? && registrations.length > competition.competitor_limit
+      raise "The given file includes #{registrations.length} accepted #{"registration".pluralize(registrations.length)}"\
+            ", while #{competition.competitor_limit} is the competitor limit."
```
Comment (jfly) [inline WcaOnRails/app/controllers/registrations_controller.rb]: nit: Maybe clearer to say ", which is more than the competitor limit of #{....}."?

File: WcaOnRails/config/locales/en.yml [hunk 11]
```diff
@@ -953,9 +953,9 @@ en:
     registration_requirements: "Registration requirements for the competition:"
     import:
       title: "Import registrations for %{comp}"
-      warning: "This action is not reversible, make sure to do this once the registration is closed."
+      warning: "This action is not reversible and should be done only once, so don't do this until after registration is closed."
```
Comment (jfly) [inline WcaOnRails/config/locales/en.yml]: If I'm understanding correctly, doesn't the code only allow this to happen once? If so, then I think we should say "and can only be done once".

However, as mentioned elsewhere, I think we should allow people to re-import. Even if they have to do something like manually delete every registration to free things up for re-importing.

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Ah, sorry I didn't explain my suggestion very well. I do agree that it's weird for the User model to know about the CSV row hash. I was picturing something more like:

```ruby
### WcaOnRails/app/controllers/registrations_controller.rb
def self.create_locked_account!(registration_csv_data)
  User.new_locked_account(
    name: registration_csv_data[:name],
    email: registration_csv_data[:email],
    wca_id: registration_csv_data[:wca_id],
    country_iso2: Country.c_find(registration_csv_data[:country]).iso2,
    gender: registration_csv_data[:gender],
    dob: registration_csv_data[:birth_date],
  ).save!
end

### models/user.rb
def self.new_locked_account(attributes = {})
  # A locked account is just a user with an empty password. It's impossible to sign into an account with an empty password, so the only way to log into a locked account is to reset its password.
  User.new(attributes.merge(encrypted_password: "")).tap do |user|
    user.define_singleton_method(:password_required?) { false }
    user.skip_confirmation!
  end
end
```

This may feel kind of boilerplate-y, but the thing I like about this is that it separates the "how do you create an account with no password?" from the "what data should go into this locked account?"

[the remark this answers is not in the document]
Comment (jfly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: I don't feel strongly about the `define_singleton_method` vs `attr_writer` approach. Maybe would be nice to link to the SO thread, but I also don't feel strongly about that. Whatever you think is best!

**Comment by jfly** [comment]
LGTM! :shipit: As I mentioned a few times, I think it would be nice for people to have some way to re-import registrations if something goes wrong, but I'm also ok with us adding support for that later if you prefer!

One difference between groupifier and wca-live is that wca-live needs to somehow handle at the door registration. We should talk to @ronaldmansveld about how he envisions handling that for wca-live. I don't know if he's planning to build a "add a new competitor" UI to wca-live.

**Comment by jonatanklosko** [comment]
I've just added an additional validation that detects when a new user is created with an email belonging to some locked account. That's how it looks one the sign up page:
image
Any wording suggestions?

**Comment by jfly** [comment]
Maybe reword the first sentence to get rid of the word "locked"? Something like "an account already exists with this email. It's been created..."

**Comment by jonatanklosko** [comment]
Done!
image
I'm gonna deploy this as soon as Travis passes then! :shipit:
